torch_frame/data/multi_embedding_tensor.py

```diff
@@ -98,29 +101,159 @@ def __getitem__(
+            >>> # list/tensor/range can select multiple rows/cols
+            >>> out.select([0, 2], dim=1)
+            MultiEmbeddingTensor(num_rows=2, num_cols=2, device='cpu')
+        """
+        dim = self._normalize_dim(dim)
+        if isinstance(index, int):
+            return self._single_index_select(index, dim=dim)
+        elif isinstance(index, slice):
+            return self._slice(index, dim=dim)
         elif isinstance(index, Tensor) and index.ndim == 1:
-            return self.index_select(index, dim=0)
-        elif isinstance(index, list):
+            return self.index_select(index, dim=dim)
+        # TODO: Don't materialize range but instead pass it to PyTorch tensor
+        # directly for possible better performance.
+        elif isinstance(index, (list, range)):
             return self.index_select(
                 torch.tensor(index, dtype=torch.long, device=self.device),
-                dim=0,
+                dim=dim,
             )
-        # TODO(akihironitta): Support more index types
         raise NotImplementedError
 
+    def narrow(
+        self,
+        dim: int,
+        start: int,
+        length: int,
+    ) -> "MultiEmbeddingTensor":
+        """Returns a narrowed version of input :class:`MultiEmbeddingTensor`.
+
+        Args:
+            dim (int): The dimension along which to narrow.
+            start (int): The starting dimension index to narrow.
+            length (int): The length of the dimension to narrow.
+
+        Returns:
+            MultiEmbeddingTensor: A :class:`MultiEmbeddingTensor` instance.
+
+        Example:
+            >>> num_rows = 2
+            >>> tensor_list = [
+            ...    torch.tensor([[0, 1, 2], [3, 4, 5]]),  # col0
+            ...    torch.tensor([[6, 7], [8, 9]]),        # col1
+            ...    torch.tensor([[10], [11]]),            # col2
+            ... ]
+            >>> out = MultiEmbeddingTensor.from_tensor_list(tensor_list)
+            >>> out
+            MultiEmbeddingTensor(num_rows=2, num_cols=3, device='cpu')
+            >>> out.narrow(dim=1, start=1, length=2)
+            MultiEmbeddingTensor(num_rows=2, num_cols=2, device='cpu')
+        """
+        dim = self._normalize_dim(dim)
+        num_data = self.num_rows if dim == 0 else self.num_cols
+        if start == 0 and start + length >= num_data:
+            return self
+        elif length <= 0:
+            return MultiEmbeddingTensor(
+                num_rows=0 if dim == 0 else self.num_rows,
+                num_cols=0 if dim == 1 else self.num_cols,
+                values=torch.tensor([], device=self.device),
+                offset=torch.tensor([0], device=self.device),
+            )
+
+        elif dim == 0:
+            return MultiEmbeddingTensor(
+                num_rows=length,
+                num_cols=self.num_cols,
+                values=self.values[start:start + length],
+                offset=self.offset,
```

Comment:
Should we also change the offset here?

Reply:
I don't think row indexing (`dim=0`) operation should change the offset, since `offset` here purely specifies the column slicing information.